catmap/api/ase_data.py

```diff
@@ -0,0 +1,1416 @@
+            fields[1] = y
+            if site_y is not None:
+                fields[-1] = site_y
+            key_y = '_'.join(fields)
+            self.formation_energies.update({key_y: Y})
+
+    def _insert_state(self, key, descriptor, slopes, intercept):
+        """ Update the formation_energy dictionary with interpolated values.
+        This is intended for use with thermodynamic scalers only.
+
+        Parameters
+        ----------
+        key : str
+            exact key of the species to be inserted self.formation_energy
+        descriptor : str or list
+            species names of descriptors.
+        slope : float or None
+        intercept : float or None
+        """
+        interpolated = intercept
+        fields = key.split('_')
+        for a in range(len(descriptor)):
+            fields[1] = descriptor[a]
+            fields[-1] = 'site'
+            energy_key = '_'.join(fields)
+            interpolated += slopes[a] * self.formation_energies[energy_key]
+        state = {key: interpolated}
+        self.formation_energies.update(state)
+
+    def _insert_frequencies(self, key, frequencies):
+        """ Update the formation_energy dictionary with interpolated values.
+        This is intended for use with thermodynamic scalers only.
+
+        Parameters
+        ----------
+        key : str
+            exact key of the species to be inserted self.freq
+        frequencies : list
+            list of frequencies.
+        slope : list
+        """
+        state = {key: frequencies}
+        self.freq.update(state)
+
+    def insert_rscaled_states(self, x, y, site_y=None,
+                              slope=None, intercept=None):
+        raise NotImplementedError("Coming in 2018.")
+
+    def make_ensemble_input_files(self, prefix, suffix, site_specific=False):
+        """ Save catmap input files for ensemble models.
+        It is advisable to use a smaller than default beef_size.
+        """
+        if self.beef_size >= 2000:
+            warnings.warn("It is advisable to use a smaller than default " +
+                          "beef_size for BEEF ensemble propagation through" +
+                          "the micro-kinetic model.")
+        raise NotImplementedError("Coming in 2018.")
+        # Create a header.
+        # headerlist = ['surface_name', 'phase', 'site_name',
+        #              'species_name', 'formation_energy',
+        #              'frequencies', 'reference', 'coverage', 'std']
+        # header = '\t'.join(headerlist)
+
+    def db_attach_formation_energy(self, fname, key_name):
+        """ Update a database file to append formation energies.
+
+        Parameters
+        ----------
+            fname : str
+                path and filename of ase database.
+        """
+        c = ase.db.connect(fname)
+        for key in tqdm(list(self.formation_energies)):
+            if 'gas' not in str(key):
+                kvp = {key_name: float(self.formation_energies[key])}
```

Comment:
I think dictionary expressions won't in python2.7. Has this been tested? Can we avoid it for now?